### PR TITLE
Github comments

### PR DESCRIPTION
**Pull request by CZYK**
# Story
At the moment our team uses functionality to easily comment the deploy-status into the pull requests. This code is triggered on every change (commit, labels, etc.) and posts the latest deploy status to the pull request. All prior messages from this user (dev-ops-user) are deleted to prevent duplicated messages.

This works perfectly fine. Now that the amount of products, which needs this functionality, starts expanding it's time to abstract these functions into an NPM package. It should start very simple with only the most important functions exposed publically.

## Definition of done
This increment is considered done when the code can be re-used across different products/teams. The package is public and may therefore not contain any sensitive data of business logic. It is fine when the package is published as an initial `RC` version so it is not yet visible on [npmjs](https://www.npmjs.com/). Of course, it should be the end-goal to publish a stable version but as an intermediate step, an RC version will do.

- [x] [RC version](https://semver.org/) `1.0.0-rc.x` published on [npmjs](https://www.npmjs.com/package/@studyportals/github)
- [x] Used in at least two implementations
- [x] _Final version `1.0.0` publised  on [npmjs](https://www.npmjs.com/)_

# Technical information for the Developers
_Two tablespoons of cinnamon and two or three egg whites_

```node
console.log(`In hindsight, ${quote}`)
```

## Tasks
- [x] NPM Package
  - [x] License
  - [x] Author
  - [x] Contributors
- [x] TypeScript
- [x] TravisCI integration
- [x] CodeClimate integration
- [x] Can comment on a Pull Request
- [x] Can delete a specific comment
- [x] Can retreive all comments
- [x] Have proper code completing (include typedefs)
- [x] Tests
- [x] Remove debug code
- [x] At least one approved review
- [x] Publish automatically into the registry after tagged commit [builds/339096253](https://travis-ci.org/studyportals/GitHub/builds/339096253)
  - [x] https://docs.travis-ci.com/user/deployment/npm/
- [x] Add pre-commit hook
- [x] update README.md
- [x] Write unit-tests with @andrei-dobre

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,57 @@
+{
+  "compilerOptions": {
+    /* Basic Options */
```
Comment (CZYK) [inline tsconfig.json]: This file should be cleaned up before merging into develop

File: tslint.json [hunk 2]
```diff
@@ -0,0 +1,11 @@
+{
+    "defaultSeverity": "error",
+    "extends": [
+        "tslint:recommended"
+    ],
+    "jsRules": {},
```
Comment (CZYK) [inline tslint.json]: Can we ditch this line?

File: tslint.json [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+    "rules": {
+        "indent": [true, "tabs"]
+    },
+    "rulesDirectory": []
```
Comment (CZYK) [inline tslint.json]: Can we ditch this line?

**Comment by CZYK** [comment]
Test case xxx

**Comment by CZYK** [comment]
Test case xxx

**Comment by CZYK** [comment]
Test case xxx

**Comment by CZYK** [comment]
Test case xxx

**Comment by CZYK** [comment]
Test case xxx

**Comment by CZYK** [comment]
Test case xxx

**Comment by CZYK** [comment]
Test case xxx

**Comment by CZYK** [comment]
Test case xxx

**Comment by CZYK** [comment]
Test case xxx

**Comment by CZYK** [comment]
Test case xxx

**Comment by CZYK** [comment]
Test case xxx

**Comment by CZYK** [comment]
Test case xxx

**Comment by CZYK** [comment]
Test case xxx

**Comment by CZYK** [comment]
Test case xxx

**Comment by CZYK** [comment]
Test case xxx

**Comment by CZYK** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by CZYK** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by CZYK** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

File: src/GitHub.ts [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+	}
+
+	public getComments(owner: string, repository: string, issueNumber?: string): Promise<any> {
+
+		const commentsUrl = this.getCommentsUrl(owner, repository, issueNumber);
+		return fetch(commentsUrl, {
+			headers: this.getHeaders(),
+			method: "GET"})
+			.then((response) => {
+				return response.json();
```
Comment (stefanhaneveer) [inline src/GitHub.ts]: Fetch does not by default only return a successfull response in here
https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch
check the "Checking that the fetch was successful" section. Goes for all fetch calls.

File: src/GitHub.ts [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+		return url;
+	}
+
+	private getCommentsUrl(owner: string, repository: string, issueNumber?: string): string {
+		return this.getIssuesUrl(owner, repository, issueNumber) + "/comments";
+	}
+
+	private getHeaders(): {} {
+
+		// TODO: using `any` is a poor mans solution and should be solved...
```
Comment (stefanhaneveer) [inline src/GitHub.ts]: I believe there is a Header interface with node-fetch

File: src/IGitHub.ts [hunk 6]
```diff
@@ -0,0 +1,5 @@
+export interface IGitHub {
+	getComments(owner: string, repository: string, issueNumber?: string): Promise<any>;
+	deleteComment(owner: string, repository: string, commentId: string): Promise<any>;
+	postComment(owner: string, repository: string, issueNumber: string, body: string): Promise<any>;
```
Comment (stefanhaneveer) [inline src/IGitHub.ts]: The official github api package was working on proper interfaces for responses rather then <any> maybe those could sooner or later separately be used if they make a separate package out of those.

File: src/test/GitHub.test.ts [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+	let GH;
+	let commentId;
+
+	before(() => {
+		GH = new GitHub(process.env.GITHUB_TOKEN);
+	});
+
+	it("Should create a comment", () => {
+
+		return GH.postComment("studyportals", "github", "1", "Test case xxx")
```
Comment (stefanhaneveer) [inline src/test/GitHub.test.ts]: These outgoing requests should be mocked for unit/integration tests.

File: src/IGitHub.ts (hunk 6 above)
Comment (CZYK) [reply src/IGitHub.ts]: That would be awesome.

File: src/test/GitHub.test.ts (hunk 7 above)
Comment (CZYK) [reply src/test/GitHub.test.ts]: Yep! Scheduled a meeting with you :)

**Comment by CZYK** [comment]
Test case xxx

**Comment by CZYK** [comment]
Test case xxx

**Comment by CZYK** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by CZYK** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by CZYK** [comment]
Test case xxx

**Comment by CZYK** [comment]
Test case xxx

**Comment by CZYK** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by CZYK** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by CZYK** [comment]
Publishing from TravisCI works 👍 
```
0.01s$ echo 'ready?'
ready?
dpl.0
1.76s$ rvm $(travis_internal_ruby) --fuzzy do ruby -S gem install dpl
Successfully installed dpl-1.8.47
1 gem installed
15.63s
dpl.1
Installing deploy dependencies
dpl.2
Preparing deploy
NPM version: 5.6.0
Authenticated with email southpark@studyportals.eu
Cleaning up git repository with `git stash --all`. If you need build artifacts for deployment, set `deploy.skip_cleanup: true`. See https://docs.travis-ci.com/user/deployment/#Uploading-Files.
Saved working directory and index state WIP on (no branch): d91cb4f Publish for my branch (testing purpose)
NPM API key format changed recently. If your deployment fails, check your API key in ~/.npmrc.
http://docs.travis-ci.com/user/deployment/npm/
~/.npmrc size: 48
dpl.3
Deploying application
+ @studyportals/github@1.0.0-alpha.7
Already up to date!
HEAD detached at d91cb4f
nothing to commit, working tree clean
Dropped refs/stash@{0} (7ab7fa3e7966d745a8a8734ebafb65e92f0f3eaf)
after_deploy
0.01s$ echo 'deployed :-)'
deployed :-)
```

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by CZYK** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by CZYK** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by CZYK** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by StudyPortalsDevOps** [comment]
Test case xxx

**Comment by CZYK** [comment]
@stefanhaneveer, together with @andrei-dobre we decided to not test our integration layer with GitHub but test our business logic instead with TDD.

**Pull request merged**